packages/protocol/contracts/L1/libs/LibAuction.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: MIT
+//  _____     _ _         _         _
+// |_   _|_ _(_) |_____  | |   __ _| |__ ___
+//   | |/ _` | | / / _ \ | |__/ _` | '_ (_-<
+//   |_|\__,_|_|_\_\___/ |____\__,_|_.__/__/
+
+pragma solidity ^0.8.18;
+
+import {AddressResolver} from "../../common/AddressResolver.sol";
+import {LibAddress} from "../../libs/LibAddress.sol";
+import {LibUtils} from "./LibUtils.sol";
+import {TaikoData} from "../../L1/TaikoData.sol";
+import {TaikoToken} from "../TaikoToken.sol";
+
+library LibAuction {
+    using LibAddress for address;
+
+    event Bid(
+        uint256 indexed id,
+        address claimer,
+        uint256 claimedAt,
+        uint256 deposit,
+        uint256 minFeePerGasInWei
+    );
+
+    event BidDepositRefunded(
+        uint256 indexed batchStartId, address claimer, uint256 refundedAt, uint256 refund
+    );
+
+    error L1_BLOCK_ID();
+    error L1_BID_AUCTION_CLOSED();
+    error L1_BID_NOT_ACCEPTABLE();
+    error L1_ID_NOT_START_OF_A_BATCH();
+
+    function bidForBatch(
+        TaikoData.State storage state,
+        AddressResolver resolver,
+        TaikoData.Config memory config,
+        uint256 batchStartBlockId,
+        uint256 minFeePerGasInWei
+    ) internal {
+        // If it verified already -> Revert, otherwise it would be possible
```

Comment:
The block ID => batch mapping should consider the fact the genesis block is at 0, and the first batch will cover block #1 to block #100, assuming batch size is 100.

```solidity
funciton blockIdToBatch(uint blockId) {
   return (blockId - 1)/batch_size;
}
```

Reply:
Block proposal and batch auction can be entirely independent: we can have a million blocks proposed and we are still auctioning the first 100K blocks, or we can have more block batch auctioned but a few blocks proposed. Decoupling these two will make things easier.

Since proofs cannot be submitted without its batch auction complete. When someone bid for a batch, we do not need to check if lastVerifiedBlockId is smaller/larger than the first block in the batch. We just need to check if the auction has started/ended.

Reply:
Lets expect provers to have their own ways of distributing blocks of a batch to its workers. They may have their own auction offchain.